Reworked sending

JarvisCraft opened the pull request:
* Renamed to MessengerAdvancement as AdvancementAPI seems to be a better name for Singleton
+ Added .grant() and .revoke() methods
+ Added default values for variables
+ Added String-to-FrameType variants of input (supporting random values)
+ Returned setters without return type
+ Added @SuppressWarnings("unchecked") for .getJSON() method
- Deprecated .save() methods

yannicklamprecht commented on AdvancementAPI.java:
```diff
@@ -71,66 +95,110 @@ public String getTitle() {
+    public MessengerAdvancement withFrame(String frame) {
+        if (frame.equalsIgnoreCase("random")) {
+            FrameType[] frames = FrameType.values();
+            setFrame(frames[(int)(Math.random()*(frames.length-1))]);
```
move `frames[(int)(Math.random()*(frames.length-1))]` to FrameType. Then you're able to do FrameType.random()

yannicklamprecht commented on AdvancementAPI.java:
```diff
@@ -71,66 +95,110 @@ public String getTitle() {
+    public void setFrame(String frame) {
+        if (frame.equalsIgnoreCase("random")) {
+            FrameType[] frames = FrameType.values();
+            setFrame(frames[(int)(Math.random()*(frames.length-1))]);
```
same here -> setFrame(FrameType.random());

yannicklamprecht commented on AdvancementAPI.java:
```diff
@@ -13,27 +15,37 @@
+    }
+
+    public MessengerAdvancement(String id) {
+        this.id = new NamespacedKey(ContentMakersPlugin.getInstance(), id);
```
Won't work this way. You've to pass the the JavaPlugin. `ContentMakersPlugin.getInstance()` only works with your plugin.

JarvisCraft replied:
Here my bad
Forgotten to change back to norm

JarvisCraft replied:
Will fix in a few minutes (when I'm from PC)

yannicklamprecht commented on AdvancementAPI.java:
```diff
@@ -44,25 +56,37 @@ public String getIcon() {
         return this;
     }
 
+    public void setIcon(String icon) {
```
Why adding this method? `withIcon(String icon)` does the work!

JarvisCraft replied:
About setters explained above - just to make the Api fit in cases when it's not necessary to have return
Though it's a thing to discuss

yannicklamprecht commented on AdvancementAPI.java:
```diff
@@ -44,25 +56,37 @@ public String getIcon() {
         return this;
     }
 
+    public void setDescription(String description) {
```
The same here. `withDescription(String description)` does it.

yannicklamprecht commented on AdvancementAPI.java:
```diff
@@ -44,25 +56,37 @@ public String getIcon() {
     }
 
-    public AdvancementAPI withBackground(String url) {
+    public void setBackground(String background) {
```
Why renaming it? The name `withBackground` has a reason. It implies thats make use of the BuilderPattern.

JarvisCraft commented:
Will update to 1.2 soon fixing issues with wrong Plugin and Moving *random* to enum

JarvisCraft commented:
As said before:
* replaced ContentMakers with JavaPlugin (plus changed this constructor)
+ made `RANDOM()` into enum inside of `FrameType`

Thanks to @ysl3000 

//PS setters are still a thing to discuss as .with<NAME>'s don't seem to fit in those cases where value is only set for one param. I suggest:
* making smth. like `.with<NAME>() { .set<NAME>; return(this); }`
OR simply having only .<NAME>()'s
//PPS I'm also planning to add the `Map<String, AdvancementAPI>` (mentioned in Spigot discussion) as static member here and (probably) add "safe-show"

ysl3000 commented:
Do you know Lombok ? I tested how good the Builder is working for our problem and it is working really well. https://github.com/Craftstuebchen/AdvancementAPI/blob/master/src/test/java/AdvancementAPITest.java

https://github.com/Craftstuebchen/AdvancementAPI/blob/master/src/main/java/io/chazza/advancementapi/AdvancementAPI.java

yannicklamprecht commented on AdvancementAPI.java:
```diff
@@ -175,51 +242,112 @@ public String getJSON() {
+    public MessengerAdvancement show(Player... players) {
+        add();
+        grant(players);
+        Bukkit.getScheduler().runTaskLater(ContentMakersPlugin.getInstance(), new Runnable() {
```
Hey I found the same problem here. `ContentMakersPlugin.getInstance()`

JarvisCraft replied:
LoL
One minute

JarvisCraft commented:
+*Fixed .send() 

Gonna work on shortening tomorrow

//Sorry, it's 1:58 in Moscow now

JarvisCraft commented:
+ Added static `List<AdvancementAPI>` to hold all built Advancements generated during the following session
+ Added two static and one non-static `.build()` methods to replace all constructor-variants which also add AdvancementAPI-instances into static list
+ Added `.unbuild()` method to remove AdvancementAPI-instances from static list
+ Overwritten `.toString()` method`
+ Edited top JavaDocs (sorry if any mistakes)
* Renamed back into AdvancementAPI
* Renamed `.with<NAME>` methods into `.<NAME>` methods
- Removed AdvancementAPI(BukkitPlugin plugin, String name) constructor and made another one private
- Removed package
- Removed setters
- Removed unnecessary debug messages

JarvisCraft commented:
The new way to create a custom Advancement to better fit the OneClass-BuilderPattern:
`//In the given example I am creating a new Advancement-Announcement which will have a Unique ID "<PLUGINNAME>:announcement_breaker", will be shown once (to players player1 and player2), will have the diamond axe as an icon and will have a bold dark_red description "BREAK'EM'ALL":
AdvancementAPI.build(plugin, "announcement_breaker")
                .title("§4§lBREAK'EM'ALL")
                .icon("minecraft:diamond_axe")
                .show(player1, player2)
                .unbuild();
//.unbuild() in the end also returns AdvancementAPI so you can use it again (though any changes in it require server restart/reload new Unique ID)`
